shell/shell.js

```diff
@@ -16,22 +16,153 @@
+  this.registered = [];
+}
+Shortcuts.prototype.register = function(acceleratorName, callback) {
+  this.registered.push([acceleratorName, callback]);
+};
+Shortcuts.prototype.testEvent = function(event, acceleratorName) {
+  if (!(acceleratorName in this.accelerators)) {
+    return false;
+  }
+  return this.accelerators[acceleratorName](event);
+};
+Shortcuts.prototype.handleEvent = function(event) {
+  for (let idx = 0; idx < this.registered.length; idx++) {
+    let [ acceleratorName, callback ] = this.registered[idx];
+    if (this.testEvent(event, acceleratorName)) {
+      callback();
+    }
+  }
+};
+
+window.addEventListener('load', () => {
+  UI.init();
+}, { once: true });
 
-  browser.loadURI(url, null, null);
-  // dump instead of console.log to write to stdout for tests.
-  dump(`opened ${url} in new window\n`);
-  Runtime.openDevTools(browser);
+window.addEventListener('unload', () => {
+  UI.destroy();
+}, { once: true });
 
-  browser.addEventListener('keydown', event => {
-    // Reload the web page when the F5 key is pressed.
-    if (event.keyCode && event.keyCode === 116 && !event.shiftKey &&
-        !event.altKey && !event.ctrlKey && !event.metaKey) {
+const UI = {
+  init: () => {
+    const browser = UI.browser = document.getElementById('content');
+    const url = UI.url = window.arguments[0];
+    const shortcuts = UI.shortcuts = new Shortcuts();
+
+    // Focus the browser window when the application is opened.
+    browser.focus();
+
+    browser.loadURI(url, null, null);
+
+    // Dump instead of log to write to stdout for tests.
+    dump(`opened ${url} in new window\n`);
+
+    // Hard-reload the web page.
+    // Windows/Linux: `Shift + F5` or `Ctrl + Shift + R`.
+    // Mac:           `Shift + F5` or `Cmd + Shift + R`.
+    shortcuts.register('HardReload', () => {
+      // Bypass proxy and cache.
+      const reloadFlags = Ci.nsIWebNavigation.LOAD_FLAGS_BYPASS_PROXY |
+        Ci.nsIWebNavigation.LOAD_FLAGS_BYPASS_CACHE;
+      browser.webNavigation.reload(reloadFlags);
+    });
+
+    // Reload the web page.
+    // Windows/Linux: `F5` or `Ctrl + R`.
+    // Mac:           `F5` or `Cmd + R`.
+    shortcuts.register('Reload', () => {
       browser.reload();
-    }
-  }, false, true);
+    });
+
+    // Toggle the DevTools.
+    // Windows/Linux: `Ctrl + Shift + I`.
+    // Mac:           `Cmd + Shift + I`.
+    shortcuts.register('ToggleDevTools', () => {
+      const toolsWindow = Runtime.toggleDevTools(browser);
+      if (toolsWindow) {
+        // DevTools window was created for the first time.
+        toolsWindow.addEventListener('keydown', onToolsKeydown);
+        toolsWindow.addEventListener('unload', onToolsUnload, { once: true });
+      }
+    });
+    const onToolsKeydown = event => {
+      // NB: the DevTools window handles its own reload key events,
+      // so we don't need to handle them ourselves.
+      // TODO: make this DRY, so we're not repeating ourselves below.
```

Comment:
Nit: the code that was "below" this comment is now above it. It also isn't as repetitive, now that the reload handlers aren't needed. So I'm not sure that the comment is necessary anymore.